h/h_api/bulk_api/command_batch.py

```diff
@@ -0,0 +1,88 @@
+"""Batching and dispatching jobs to work on."""
+
+from contextlib import contextmanager
+
+from h.h_api.exceptions import CommandSequenceException
+
+
+class CommandBatch:
+    """A batch of commands.
+
+    This object will group similar commands together and decide when they
+    should be executed. The number of commands should always be a number we
+    are comfortable dealing with in memory at once.
+    """
```

Comment:
Ok, so you just keep calling `add()` and `CommandBatch` will call the given `on_flush()` after each 100 adds (by default) or when the type of command you're adding changes? And I'm guessing `on_flush` is going to be some kind of method for executing a batch of commands?

Reply:
Yes that's right. I think there could conceivably be more complication than that in a general solution, as upsert commands on the same object with different queries could be incompatible, but this works for now.